### ISSUE TITLE
Calling `toString` on JmlCompilationUnit creates incorrect output for quantified expressions

### ISSUE THREAD

**Issue by Thiez** [OpenJML/OpenJML#525]
The problem occurs with OpenJML 0.8.13, and at least one earlier version (I lost the original download, but the version reports as "openjml OpenJML-20170315".

The following program demonstrates the problem:

```java
import org.jmlspecs.openjml.*;

public class Example {

    /*@
      ensures \forall int i; i = 5; i < (i + 1);
      ensures \exists int i; i = 5; i < (i + 1);
    */
    public static void main(String[] args) throws Exception {
        IAPI api = Factory.makeAPI("-jml", "-strictJML");
        System.out.println(api.parseSingleFile("./Example.java").toString());
    }

}
```

When run with `javac Example.java -cp openjml.jar && java -cp '.:openjml.jar' Example` it produces the following output:

```java
import org.jmlspecs.openjml.*;

public class Example {
    /*@
      ensures \forall int i;; i = 5; i < (i + 1); 
      ensures \exists int i;; i = 5; i < (i + 1); 
   */

  public static void main(String[] args) throws Exception {
    IAPI api = Factory.makeAPI("-jml", "-strictJML");
    System.out.println(api.parseSingleFile("./Example.java").toString());
  }
}
```

As can be seen, the pretty printing has inserted additional ';' characters after the variable declarations in the `\forall` and `\exists` expressions. The problem *appears* to be related to the `int i;` being a `JCVariableDecl`, which will print a ';' as part of its own pretty-printing, but `JmlPretty.visitJmlQuantifiedExpr` will *also* print a `;` after printing the variable declarations.

This also goes very wrong with something like `ensures \forall int i, j; i = 5 && j == 6; i < j;`, which will pretty print like `ensures \forall int i;, int j;; i = 5 && j == 6; i < j;`.

**Comment by davidcok** [comment]
OK. Thanks for the example — that is incorrect and easily fixed.

**Comment by Thiez** [comment]
I got it working working locally by overriding `visitJmlQuantifiedExpr` with the following implementation:

```java
public void visitJmlQuantifiedExpr(JmlTree.JmlQuantifiedExpr that) {
	try {
		print(that.op.internedName());
		print(" ");
		boolean first = true;
		Writer savedOut = out;
		try {
			StringBuffer buffer;
			{
				StringWriter tempOut = new StringWriter();
				buffer = tempOut.getBuffer();
				this.out = tempOut;
			}

			for (JCTree.JCVariableDecl n : that.decls) {
				if (!first) {
					print(", ");
				} else {
					first = false;
				}

				n.accept(this);
				out.flush();
				int newBufSize = buffer.length();
				if (newBufSize > 0 && buffer.charAt(newBufSize - 1) == ';') {
					buffer.deleteCharAt(newBufSize - 1);
				}
			}

			out.flush();
			savedOut.append(buffer.toString());
		} finally {
			out = savedOut;
		}

		print("; ");
		if (that.range != null) {
			that.range.accept(this);
		}
		print("; ");
		if (that.value != null) {
			that.value.accept(this);
		} else {
			print("????:");
		}
	} catch (IOException e) {
		perr(that, e);
	}
}
```

It's a bit of a hack but it avoids modifying the logic for pretty printing variable declarations, and appears to work correctly for my purposes.

**Comment by davidcok** [comment]
The ‘correct’ solution is to set the precedence correctly - which will be in the next release.  Let me know if it is critical for you, given this solution.

**Comment by Thiez** [comment]
I can wait for the next release. Thank you for the quick fix!

**Comment by davidcok** [comment]
It is in the current release that I put out on the weekend.

**Comment by Thiez** [comment]
This is still broken for me in loop invariants. New example code:

```java
import org.jmlspecs.openjml.*;

public class Example {
    /*@
      ensures \forall int i, j; i = 5 && j == 6; i < j;
      ensures \exists int i; i = 5; i < (i + 1);
    */
    public static void main(String[] args) throws Exception {
        //@ loop_invariant \forall int j; j > 100; i < j;
        for (int i = 0; i < 5; i++);
        IAPI api = Factory.makeAPI("-jml");
        System.out.println(api.parseSingleFile("./Example.java").toString());
    }
}
```

This prints:

```java
import org.jmlspecs.openjml.*;

public class Example {
    /*@
      ensures \forall int i, int j; i = 5 && j == 6; i < j; 
      ensures \exists int i; i = 5; i < (i + 1); 
   */

  public static void main(String[] args) throws Exception {
    //@ loop_invariant \forall int j;; j > 100; i < j;
    for (int i = 0; i < 5; i++) ;
    IAPI api = Factory.makeAPI("-jml");
    System.out.println(api.parseSingleFile("./Example.java").toString());
  }
}
```

**Comment by davidcok** [comment]
Fixed in 0.8.16
